docsTech/agnostic-styling.md

```diff
@@ -0,0 +1,133 @@
+  zIndex: 1,
+  opacity: ownerState.expanded ? 0 : 1,
+  pointerEvents: ownerState.expanded ? 'none' : 'auto',
+  transition: vars.transition(['opacity']),
+}));
+
+// usage
+const classes = useUtilityClasses(rootProps)
+<GridQuickFilterTrigger
+  ownerState={{ ...rootProps, expanded: true }}
+  className={classes.trigger}
+/>
+```
+
+#### Proposed syntax
+
+```javascript.jsx
+const triggerSlot = slot({
+  name: 'MuiDataGrid',
+  slot: 'toolbarQuickFilterTrigger',
+  as: ToolbarButton,
+}, {
+  root: {
+    zIndex: 1,
+    opacity: 1,
+    pointerEvents: 'auto',
+    transition: vars.transition(['opacity']),
+  },
+  expanded: {
+    opacity: 0,
+    pointerEvents: 'none',
+  },
+}));
+
+// usage
+const trigger = useGridStyleSlot(rootProps, triggerSlot)
+<trigger.component
+  ownerState={rootProps}
+  className={clsx(
+    trigger.classes.root,
+    expanded && trigger.classes.expanded,
+  )}
+/>
+```
+
+#### Proposed compilation (emotion/styled)
+
+```javascript.jsx
+const triggerSlot = {
+  component: styled(ToolbarButton, {
+    name: 'MuiDataGrid',
+    slot: 'toolbarQuickFilterTrigger',
+  })({
+    /* root: */
+    zIndex: 1,
+    opacity: 1,
+    pointerEvents: 'auto',
+    transition: vars.transition(['opacity']),
+
+    /* expanded: */
+    '&.MuiDataGrid-toolbarQuickFilterTrigger--expanded': {
+      opacity: 0,
+      pointerEvents: 'none',
+    },
+  }),
+  classes: {
+    root: 'MuiDataGrid-toolbarQuickFilterTrigger',
+    expanded: 'MuiDataGrid-toolbarQuickFilterTrigger--expanded',
+  },
+}
+
+// usage
+const trigger = useGridStyleSlot(rootProps, triggerSlot)
+<trigger.component
+  ownerState={rootProps}
+  className={clsx(
+    trigger.classes.root,
+    expanded && trigger.classes.expanded,
+  )}
+/>
+```
+
+#### Proposed compilation (no emotion)
+
+```javascript.jsx
+import './[currentFilename].autogenerated.css';
+
+const triggerSlot = slot({
+  name: 'MuiDataGrid',
+  slot: 'toolbarQuickFilterTrigger',
+  as: ToolbarButton,
+}, {
+  root: 'MuiDataGrid-toolbarQuickFilterTrigger',
+  expanded: 'MuiDataGrid-toolbarQuickFilterTrigger--expanded',
+}))
+
+// usage
+const trigger = useGridStyleSlot(rootProps, triggerSlot)
+<trigger.component
+  ownerState={rootProps}
+  className={clsx(
+    trigger.classes.root,
+    expanded && trigger.classes.expanded,
+  )}
+/>
+```
+
+#### Downsides
+
+- Right now, the development mode uses the packages source code directly to run. If we require a build step to turn the core grid codebase into emotion/non-emotion builds, the DX for working on the material grid might become worst. Keep in mind that in this scenario, most of the datagrid codebase would be inside the core grid package(s), while only leaf material code would live inside the `@mui/x-data-grid{,pro,premium}` packages.
+
+- Breaking change: users can't rely on non-standard `ownerState` values (like `.expanded` above) to add overrides, we'd likely standardize `ownerState` to just `rootProps`.
```

Comment:
Is this related to the emotion/non-emotion builds at all?
Or is it an unrelated breaking change that we did not anticipate and discovered after releasing the v8 stable?

Reply:
> I guess we could leave current non-standard ownerState uses as-is to avoid breaking changes, but we need to handle conditional styles with classnames anyway.

If [this](https://github.com/mui/mui-x/blob/d689f32c8ddbacdf04bcbcb36f25ba8b8524d056/packages/x-data-grid/src/components/toolbar/GridToolbarQuickFilter.tsx#L77-L85) is non-standard, then this suggestion seems like a good path forward.

Reply:
I've never worked much with material-ui, so my understanding is limited to what Lukas said [in slack](https://mui-org.slack.com/archives/C011VC970AW/p1745216072553609?thread_ts=1745143638.249159&cid=C011VC970AW). But how precisely should the classnames, conditional styles & overrides resolvers work is still fuzzy for me. There doesn't seem to be a strictly enforced convention, so the overall result is messy.

Reply:
> I guess we could leave current non-standard ownerState uses as-is to avoid breaking changes, but we need to handle conditional styles with classnames anyway.

So there are no breaking changes that are blocking the agnostic styling effort?

Reply:
Nothing blocking.

Reply:
Actually, now that I think about it, the bulletpoint after this one about modifier classnames would be a breaking change. Some styles that are currently applied via emotion would gain more specificity, so users overriding them via `sx` or `styleOverrides` could have their styles broken.